extensions/commands/qt/cmd_upgrade_qt_recipe.py

```diff
@@ -0,0 +1,226 @@
+    mirrors.append(link)
+    mirrors.append(f"https://download.qt.io/archive/qt/{version.major}.{version.minor}/{version}/single/{archive_name}")
+    with session.get(f"{link}.meta4") as req:
+        req.raise_for_status()
+        tree = xml.etree.ElementTree.fromstring(req.text)
+        if tree.tag != "{urn:ietf:params:xml:ns:metalink}metalink":
+            ConanOutput().error(f"meta link root tag incorrect: expected \"metalink\" but got {tree.tag}")
+            sys.exit(-1)
+        file = tree.find("{urn:ietf:params:xml:ns:metalink}file")
+        if not file:
+            ConanOutput().error(f"Could not find `file` tag in {link}.meta4 file content")
+            sys.exit(-1)
+        hash_element = file.find("{urn:ietf:params:xml:ns:metalink}hash[@type='sha-256']")
+        if hash_element is None:
+            ConanOutput().error(f"Could not find hash tag in {link}.meta4 file content")
+            sys.exit(-1)
+        sources_hash = hash_element.text
+        if sources_hash is None:
+            ConanOutput().error(f"Could not find hash text in {link}.meta4 file content")
+            sys.exit(-1)
+        mirrors.extend(node.text for node in file.findall("{urn:ietf:params:xml:ns:metalink}url") if node.text)
+    return sources_hash, mirrors
+
+
+def recipe_folder(version: Version) -> str:
+    return f"{version.major}.x.x"
+
+
+def update_conandata_yml(version: Version, sources_hash: str, mirrors: list[str]) -> None:
+
+    conan_data_yml_path = f"{recipe_folder(version)}/conandata.yml"
+
+    with open(conan_data_yml_path, "r") as conandata_file:
+        conandata = yaml.safe_load(conandata_file)
+        if str(version) in conandata["sources"]:
+            ConanOutput().error(f"version \"{version}\" already present in conandata.yml sources")
+            sys.exit(-1)
+        if str(version) in conandata["patches"]:
+            ConanOutput().error(f"version \"{version}\" already present in conandata.yml patches")
+            sys.exit(-1)
+        patches = list(conandata["patches"].values())[0]
+
+    lines = []
+    with open(conan_data_yml_path) as conandata_file:
+        sources_inserted = False
+        patches_inserted = False
+        for line in conandata_file:
+            lines.append(line)
+            if not sources_inserted and line.startswith("sources:"):
+                lines.append(f"  \"{version}\":\n")
+                lines.append("    url:\n")
+                lines.extend(f"      - \"{m}\"\n" for m in mirrors)
+                lines.append(f"    sha256: \"{sources_hash}\"\n")
+                sources_inserted = True
+
+            if not patches_inserted and line.startswith("patches:"):
+                lines.append(f"  \"{version}\":\n")
+                lines.extend(f"    {line}\n" for line in yaml.safe_dump(patches, default_style='"', default_flow_style=False).splitlines())
+                sources_inserted = True
```

Comment:
There is a little diff that is not very hard to correct manually in the `conandata.yml`:

```yaml
  "6.5.0":
    - "base_path": "qtbase/cmake"
      "patch_file": "patches/qt6-pri-helpers-fix.diff"
...
```

to :

```yaml
  "6.5.0":
    - base_path: "qtbase/cmake"
      patch_file: "patches/qt6-pri-helpers-fix.diff"
...
```
It is purely stylistic because I wonder if it is not valid YAML.

Reply:
AFAIK, it is valid yaml to quote keys. Either we don't use `default_style='"'` argument and no keys or values are quoted, either we use `default_style='"'` and both keys and values are quoted. I don't know how we could do differently